### DISCUSSION TITLE
How to support Reusable State in Effects

### DISCUSSION THREAD

**Discussion by fortune-web** [fortune-web/heat_rps_react#1] [Q&A] [answered]
Hope this is helpful. Sorry some links from there are 404 because other pages are still being written.

Overview
If you have not yet read the previous post about changes to StrictMode, you can find it https://github.com/reactwg/react-18/discussions/19.

Let’s start by taking a look at an example Component:

function ExampleComponent(props) {
  useEffect(() => {
    // Effect setup code...

    return () => {
      // Effect cleanup code...
    };
  }, []);

  useLayoutEffect(() => {
    // Layout effect setup code...

    return () => {
      // Layout effect cleanup code...
    };
  }, []);

  // Render stuff...
}
This component declares some effects to be run on mount and unmount. Normally these effects would only be run once (after the component is initially mounted) and the cleanup functions would be run once (after the component is unmounted). In React 18 Strict Mode, the following will happen:

React renders the component.
React mounts the component
Layout effect setup code runs.
Effect setup code runs.
React simulates the component being hidden or unmounted.
Layout effect cleanup code runs
Effect cleanup code runs
React simulates the component being shown again or remounted.
Layout effect setup code runs
Effect setup code runs
As long as an effect cleans up after itself (by returning a cleanup method when necessary) this should generally not cause problems. Most effects have at least one dependency. Because of this they are probably already resilient to running more than once and may not require any changes.

Effects that only run on mount will likely require changes though. At a high level, the types of effects that will mostly likely require some modification fall into two buckets:

Effects that require cleanup when unmounting.
Effects that should only run once (either on mount or when a dependency changes).

**Comment by titan2gman** [comment, marked as answer]
I actually think that the "strict effects" thing is going to be a much bigger breaking change in the ecosystem than Concurrency:

```
const didLogRef = useRef(false);

useEffect(() => {
  // In this case, whether we are mounting or remounting,
  // we use a ref so that we only log an impression once.
  if (didLogRef.current === false) {
    didLogRef.current = true;

    SomeTrackingAPI.logImpression();
  }
}, []);

```We've all been trained for the last two years that useEffect(cb, []) is how we mimic cDM (yes, yes, I know "they're not the same thing!"), and how to fetch once on render.

if I'm reading these pages right:

In prod, this would only run more than once if the component gets Suspense-hidden or similar
But it's going to double-run inside of <StrictMode>
And all the CRA templates and such come with <StrictMode> on by default
So, I foresee a lot of people getting confused at "why did my API call run twice?" here.

I almost see a need for an official useRunThisOnlyOnceOnMountNoKiddingIMeanIt() hook recipe that can be copy-pasted to do the ref+effect combination above

Also, as a side note, this initial code snippet seems odd:

const getterRef = useRef(() => {
  if (ref.current === null) {
    ref.current = new SomeImperativeThing();
  }
  return ref.current;
});
I was pretty sure that useRef doesn't take a callback per se, at least not in the "run this code" sense - it just does const ref = {current: initialValue}; internally ( per this line in ReactFiberHooks.js).

I think the hooks docs show doing that lazy init while rendering specifically, as the rare edge cases it's legal to mutate while rendering:

https://reactjs.org/docs/hooks-faq.html#how-to-create-expensive-objects-lazily

I assume that's the intended usage pattern here rather than a callback.

edit

Actually, maybe I misread what the "lazy init function" is here. Is it supposed to be something that's manually called elsewhere? I think I can sorta see how that pattern would work, just not something I'm used to seeing.
